zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -0,0 +1,102 @@
+use crate::{
+    evm_circuit::{
+        execution::bus_mapping_tmp::ExecStep,
+        param::MAX_GAS_SIZE_IN_BYTES,
+        table::{FixedTableTag, Lookup},
+        util::{
+            constraint_builder::{
+                ConstraintBuilder, StateTransition, Transition,
+            },
+            math_gadget::RangeCheckGadget,
+            Cell,
+        },
+    },
+    util::Expr,
+};
+use halo2::{
+    arithmetic::FieldExt,
+    circuit::Region,
+    plonk::{Error, Expression},
+};
+
+/// Construction of execution result that stays in the same call context, which
+/// lookups the opcode and verifies the execution result is responsible for it,
+/// then calculates the gas_cost and constrain the state transition.
+#[derive(Clone, Debug)]
+pub(crate) struct SameContextGadget<F> {
+    opcode: Cell<F>,
+    sufficient_gas_left: RangeCheckGadget<F, MAX_GAS_SIZE_IN_BYTES>,
+}
+
+impl<F: FieldExt> SameContextGadget<F> {
+    pub(crate) fn construct(
+        cb: &mut ConstraintBuilder<F>,
+        opcode: Cell<F>,
+        mut state_transition: StateTransition<F>,
+        dynamic_gas_cost: Option<Expression<F>>,
+    ) -> Self {
+        cb.opcode_lookup(opcode.expr());
+        cb.add_lookup(Lookup::Fixed {
+            tag: FixedTableTag::ResponsibleOpcode.expr(),
+            values: [
+                cb.execution_result().as_u64().expr(),
+                opcode.expr(),
+                0.expr(),
+            ],
+        });
```

Comment:
`opcode` is checked to be handled by current `execution_result` using lookup.

Reply:
I understand  `SameContextGadget` is only used for op code successful case as in failure case execution_result may be not op code itself but errors , right ?

Reply:
Yes, `ExecutionResult` it not mapping 1-1 to opcodes, it aims to flatten the possible nested errors and reduce the redundancy. And `SameContextGadget` here means the `StepState` is not going to switch to another context, and this is the case in most execution results.